### PR TITLE
Switch to scikit-build-core

### PR DESCRIPTION
**Pull request by vyasr**
## Description

Contributes to rapidsai/build-planning#2

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: dependencies.yaml [hunk 1]
```diff
@@ -62,22 +62,20 @@ dependencies:
...
         packages:
-          - &cmake_ver cmake>=3.26.4
           - cython>=3.0.0
-          - ninja
-          - scikit-build>=0.13.1
-          - tomli
       - output_types: conda
         packages:
           - c-compiler
+          - &cmake_ver cmake>=3.26.4
```
Comment (bdice) [inline dependencies.yaml:69]: Why was `cmake` removed from `[requirements, pyproject]`? We still want/need a way to pin/bound our CMake version even if `scikit-build-core` carries a transitive dependency on `cmake`.

File: python/CMakeLists.txt [hunk 2]
```diff
@@ -16,16 +16,14 @@ cmake_minimum_required(VERSION 3.26.4 FATAL_ERROR)
 
 set(rmm_version 24.02.00)
 
-include(../fetch_rapids.cmake)
+file(DOWNLOAD https://raw.githubusercontent.com/rapidsai/rapids-cmake/branch-24.02/RAPIDS.cmake
```
Comment (bdice) [inline python/CMakeLists.txt:19]: Why was this changed? We should continue to use `fetch_rapids.cmake` because it centralizes how we fetch/use rapids-cmake for both C++ and Python builds. However, if this must be changed for some reason, we'll need to be sure `update-version.sh` is updated correspondingly.

File: dependencies.yaml (hunk 1 above)
Comment (bdice) [reply dependencies.yaml:69]: It looks like `cmake` is only included in `scikit-build-core[wheels]`.

https://github.com/scikit-build/scikit-build-core/blob/565bd6b67f730acc30f982059b507c04e98fa40f/pyproject.toml#L82

File: python/pyproject.toml [hunk 3]
```diff
@@ -112,8 +107,15 @@ skip = [
     "__init__.py",
 ]
 
-[tool.setuptools]
-license-files = ["LICENSE"]
+[tool.scikit-build]
```
Comment (bdice) [inline python/pyproject.toml:110]: Can we sort the keys in this section?

File: dependencies.yaml (hunk 1 above)
Comment (vyasr) [reply dependencies.yaml:69]: scikit-build-core will conditionally install cmake as needed into the build environment. See https://scikit-build-core.readthedocs.io/en/latest/getting_started.html#python-package-configuration

File: python/CMakeLists.txt (hunk 2 above)
Comment (vyasr) [reply python/CMakeLists.txt:19]: This looks like a typo from testing. I'll revert.

File: python/pyproject.toml (hunk 3 above)
Comment (vyasr) [reply python/pyproject.toml:110]: I have no way to gauge "importance". Alphabetical is fine with me.

File: dependencies.yaml (hunk 1 above)
Comment (bdice) [reply dependencies.yaml:69]: Hmm... **(a)** I don't trust this to be correct (we regularly require specific CMake versions, or have to exclude specific versions with known bugs), and **(b)** I just don't _like_ this behavior because it makes too much of the build system "magic." I would prefer to continue specifying `cmake` pinnings (though I don't care either way about ninja).

File: conda/recipes/rmm/meta.yaml [hunk 4]
```diff
@@ -61,9 +61,7 @@ requirements:
     - cython >=3.0.0
     - librmm ={{ version }}
     - python
-    - scikit-build >=0.13.1
-    - setuptools >=61.0.0
-    - tomli  # [py<311]
+    - scikit-build-core
```
Comment (bdice) [inline conda/recipes/rmm/meta.yaml:64]: Align with the other specifications.
```suggestion
    - scikit-build-core>=0.7.0
```

File: python/pyproject.toml [hunk 5]
```diff
@@ -112,8 +107,15 @@ skip = [
...
+build-dir = "build/{wheel_tag}"
+cmake.build-type = "Release"
+cmake.minimum-version = "3.26.4"
+ninja.make-fallback = true
+sdist.reproducible = true
```
Comment (bdice) [inline python/pyproject.toml:115]: Do you know what this flag does? I read the docs but I don't fully understand its purpose. https://scikit-build-core.readthedocs.io/en/latest/api/scikit_build_core.settings.html#scikit_build_core.settings.skbuild_model.SDistSettings.reproducible

File: python/setup.py
Comment (bdice) [inline python/setup.py:1]: Good riddance, `setup.py`!

File: dependencies.yaml (hunk 1 above)
Comment (robertmaynard) [reply dependencies.yaml:69]: I agree that we need the ability to control the version resolution logic of cmake. If that can be done via toml file, that works for me

Comment (vyasr) [reply dependencies.yaml:69]: Sounds good, I can put that back then. We don't have to care about the cases where scikit-build-core is worried about wheels not being available anyway so that's OK.

File: python/pyproject.toml (hunk 5 above)
Comment (vyasr) [reply python/pyproject.toml:115]: Yeah, it basically guarantees binary equivalence between two sdists produced from the same source. In particular, it will normalize things like file permissions (in tarballs, not the underlying files) and timestamps in a way that is agnostic to the current build environment. It ensures that if you build an sdist at two different times but with identical code you get binary-equivalent sdists.

**Review by ajschmidt8 — APPROVED** [review]
